Temperature chapter (#128)
festim-dev/FESTIM-workshop: book/content/temperatures/temperatures_basic.md

---
jupyter:
  jupytext:
    formats: ipynb,md
    text_representation:
      extension: .md

>       format_name: markdown

Reviewer: None of the cells are executed if the format is not myst! In the compiled website, the final pyvista plots aren't shown.

The conversion command is `jupytext --set-format ipynb,myst my_notebook.ipynb`

      format_version: '1.3'
      jupytext_version: 1.17.3
  kernelspec:
    display_name: festim-workshop
    language: python
    name: python3
---

# Basic functionality #

Definition of a temperature field or problem is essential for hydrogen transport and FESTIM as a whole. Users can define it as a fixed value or as a function of space and time. Read more about temperature in FESTIM _[here](https://festim.readthedocs.io/en/fenicsx/userguide/temperature.html)_. FESTIM's unit for temperature is Kelvin.

This tutorial will go over the basics of defining the temperature in a FESTIM simulation.

Objectives:
* Learn how to define fixed temperature values
* Learn how to define space and time dependent temperature functions
* Learn how to define temperature as a Dolfinx expression

## Defining temperatures in FESTIM ##

The simplest way to define temperature is as a fixed value:

```python
import festim as F

my_model = F.HydrogenTransportProblem()
my_model.temperature = 300  # K 
```

> To to define a temperature function we can use ` lambda ` functions, such as this space-dependent function:

Reviewer 2: "To to" need to delete one

$$ 

T(x) = T_0 e^{-x}

$$

```python
import festim as F
import ufl

T0 = 300
my_model = F.HydrogenTransportProblem()
my_model.temperature = lambda x: T0 * ufl.exp(-x)  # K
```

For a time dependent function such as:

$$

T(t) = T_0 sin(t)

$$

```python
my_model.temperature = lambda t: T0 * ufl.sin(t)
```

For a function in $ x, y, z, t $ space:

$$

T(x,y,z,t) = T_0 cos(x)sin(t) + 5y - 2z

$$

```python
my_model.temperature = lambda x,t: T0 * ufl.cos(x[0])*ufl.sin(t) + 5*x[1] - 2*x[2]
```

<div class="alert alert-block alert-info">
When defining custom functions for values, only the arguments x, t and T can be defined. Spatial coordinates can be referred to by their indices, such as x[0], x[1], and x[2], regardless of the coordinate system used. Time dependence must use t, and T for temperature dependence.
</div>

## Defining temperature as a FEniCS expression or function ##

Users can define temperature as a Dolfinx `Function`, which can be helpful when implementing multiphysics simulations (such as in OpenFOAM or OpenMC). For example, we want to have temperature function defined as:

$$

> T(x,y) = 300 e^{((x-0.5)^2 + (y-0.5)^2)}

Reviewer 2: This grows exponentially outward. Did you perhaps intend a Gaussian shape with a negative exponent

Author: This looks a lot better, thanks!

from mpi4py import MPI
import festim as F
from basix.ufl import element

mesh_fenics = create_unit_square(MPI.COMM_WORLD, 20, 20)
my_model = F.HydrogenTransportProblem()
my_model.mesh = F.Mesh(mesh_fenics)

el = element("Lagrange", mesh_fenics.topology.cell_name(), 2)
V = dolfinx.fem.functionspace(my_model.mesh.mesh, el)

temperature = dolfinx.fem.Function(V)
x = ufl.SpatialCoordinate(temperature.function_space.mesh)[0]
y = ufl.SpatialCoordinate(temperature.function_space.mesh)[1]
interpolation = temperature.function_space.element.interpolation_points()
expr = dolfinx.fem.Expression(300*ufl.exp((x-0.5)**2 + (y-0.5)**2), interpolation)
                                
temperature.interpolate(expr)
```

```python tags=["hide-input"]
from dolfinx import plot
import pyvista

topology, cell_types, geometry = plot.vtk_mesh(V)
function_grid = pyvista.UnstructuredGrid(topology, cell_types, geometry)
function_grid.point_data["T"] = temperature.x.array.real
function_grid.set_active_scalars("T")

pyvista.start_xvfb()
pyvista.set_jupyter_backend("html")

plotter = pyvista.Plotter()
plotter.add_mesh(function_grid, cmap="inferno", show_edges=False, opacity=1)
plotter.view_xy()

> plotter.add_text("Spatially varying diffusivity", font_size=12)

Reviewer 2: The label “Spatially varying diffusivity”, but the plot is actually of temperature?

---
jupytext:
  formats: ipynb,md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.17.3
kernelspec:
  display_name: festim-workshop
  language: python
  name: python3
---

# Basic functionality #

Definition of a temperature field or problem is essential for hydrogen transport and FESTIM as a whole. Users can define it as a fixed value or as a function of space and time. Read more about temperature in FESTIM _[here](https://festim.readthedocs.io/en/fenicsx/userguide/temperature.html)_. FESTIM's unit for temperature is Kelvin.

This tutorial will go over the basics of defining the temperature in a FESTIM simulation.

Objectives:

> * Learn how to define fixed temperature values

Reviewer: ```suggestion
* Learn how to define homogeneous temperature
```

* Learn how to define space and time dependent temperature functions

> * Learn how to define temperature as a Dolfinx expression

Reviewer: ```suggestion
* Learn how to define temperature as a `dolfinx.fem.Function`
```

+++

## Defining temperatures in FESTIM ##

> The simplest way to define temperature is as a fixed value:

Reviewer: ```suggestion
The simplest way to define temperature is as a homogeneous value:
```

```{code-cell} ipython3
import festim as F

my_model = F.HydrogenTransportProblem()
my_model.temperature = 300  # K 
```

> To define a temperature function we can use ` lambda ` functions, such as this space-dependent function:

Reviewer: ```suggestion
To define a space-dependent temperature function we can use ` lambda ` functions:
```

$$ 

T(x) = T_0 e^{-x}

$$

```{code-cell} ipython3
import festim as F
import ufl

T0 = 300
my_model = F.HydrogenTransportProblem()
my_model.temperature = lambda x: T0 * ufl.exp(-x)  # K
```

> For a time dependent function such as:

Reviewer: ```suggestion
For a time-dependent function such as:
```

$$

T(t) = T_0 sin(t)

$$

```{code-cell} ipython3
my_model.temperature = lambda t: T0 * ufl.sin(t)
```

For a function in $ x, y, z, t $ space:

$$

T(x,y,z,t) = T_0 cos(x)sin(t) + 5y - 2z

$$

```{code-cell} ipython3
my_model.temperature = lambda x,t: T0 * ufl.cos(x[0])*ufl.sin(t) + 5*x[1] - 2*x[2]

> ```

Reviewer: I'm not sure this is how we implemented this in FESTIM, we might have implemented x,y,z directly instead of x[0], x[1]

A way to test this is with a complete example. FESTIM wouldn't throw an error until `initialise()`

<div class="alert alert-block alert-info">
When defining custom functions for values, only the arguments x, t and T can be defined. Spatial coordinates can be referred to by their indices, such as x[0], x[1], and x[2], regardless of the coordinate system used. Time dependence must use t, and T for temperature dependence.
</div>

+++

> ## Defining temperature as a FEniCS expression or function ##

Reviewer: Here you don't really define the temperature as an expression do you? only as a `dolfinx.fem.Function`, and you use the expression to define this Function.

I would be more explicit in the title

Users can define temperature as a Dolfinx `Function`, which can be helpful when implementing multiphysics simulations (such as in OpenFOAM or OpenMC). For example, we want to have temperature function defined as:

$$

T(x,y) = 300 e^{-((x-0.5)^2 + (y-0.5)^2)}

$$

To define the temperature as a function or expression, we can use `ufl`:

```{code-cell} ipython3
import dolfinx
from dolfinx.mesh import create_unit_square
from mpi4py import MPI
import festim as F
from basix.ufl import element

mesh_fenics = create_unit_square(MPI.COMM_WORLD, 20, 20)
my_model = F.HydrogenTransportProblem()
my_model.mesh = F.Mesh(mesh_fenics)

el = element("Lagrange", mesh_fenics.topology.cell_name(), 2)
V = dolfinx.fem.functionspace(my_model.mesh.mesh, el)

temperature = dolfinx.fem.Function(V)
x = ufl.SpatialCoordinate(temperature.function_space.mesh)[0]

> y = ufl.SpatialCoordinate(temperature.function_space.mesh)[1]

Reviewer: ```suggestion
coords = ufl.SpatialCoordinate(temperature.function_space.mesh)
x = coords[0]
y = coords[1]
```

  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.17.3
kernelspec:
  display_name: festim-workshop
  language: python
  name: python3
---

# Basic functionality #

Definition of a temperature field or problem is essential for hydrogen transport and FESTIM as a whole. Users can define it as a fixed value or as a function of space and time. Read more about temperature in FESTIM _[here](https://festim.readthedocs.io/en/fenicsx/userguide/temperature.html)_. FESTIM's unit for temperature is Kelvin.

This tutorial will go over the basics of defining the temperature in a FESTIM simulation.

Objectives:
* Learn how to define homogeneous temperature
* Learn how to define space and time dependent temperature functions
* Learn how to define temperature as a `dolfinx.fem.Function`

+++

## Defining temperatures in FESTIM ##

The simplest way to define temperature is as a homogeneous value:

```{code-cell} ipython3
import festim as F

my_model = F.HydrogenTransportProblem()
my_model.temperature = 300  # K 
```

To define a space-dependent temperature function we can use ` lambda ` functions:

$$ 

T(x) = T_0 e^{-x}

> $$

Reviewer 3: Might it be possible to use pure latex here?

i.e.

```
\begin{equation}
    T(x) = T_0 \exp(-x)
\end{equation}
```

```{code-cell} ipython3
import festim as F
import ufl

T0 = 300
my_model = F.HydrogenTransportProblem()
my_model.temperature = lambda x: T0 * ufl.exp(-x)  # K
```

For a time-dependent function such as:

$$

T(t) = T_0 sin(t)

$$

```{code-cell} ipython3
my_model.temperature = lambda t: T0 * ufl.sin(t)
```

For a function in $ x, y, z, t $ space:

$$

T(x,y,z,t) = T_0 cos(x)sin(t) + 5y - 2z

> $$

Reviewer 3: Fr some reason the cos and sin are italics, do they need a backslash beforehand to render properly?

```{code-cell} ipython3
my_model.temperature = lambda x,t: T0 * ufl.cos(x[0])*ufl.sin(t) + 5*x[1] - 2*x[2]

> ```

Reviewer 3: I might be tempted to remove this 3rd example all together and integrate something similar into the example below

Author: changed it to spatial dependence only, kept the example

<div class="alert alert-block alert-info">

> When defining custom functions for values, only the arguments x, t and T can be defined. Spatial coordinates can be referred to by their indices, such as x[0], x[1], and x[2], regardless of the coordinate system used. Time dependence must use t, and T for temperature dependence.

Reviewer 3: Wouldn't worry too much about the T dependence in this note, at least, can't have a temperature-dependent temperature.

Author: would it be helpful since it points out the different between T and t?

> </div>

Reviewer 3: ```suggestion
```{note} Note
When defining custom functions for values, only the arguments x, t and T can be defined. Spatial coordinates can be referred to by their indices, such as x[0], x[1], and x[2], regardless of the coordinate system used. Time dependence must use t, and T for temperature dependence.
```
```

Reviewer 3: This would fit in better with other pages as a note instead

+++

We can assign a unit cube mesh and run the simulation to see how a spatially-dependent temperature field affects the transport:

```{code-cell} ipython3
from dolfinx.mesh import create_unit_cube
from mpi4py import MPI
import ufl
import festim as F
import numpy as np

mesh = F.Mesh(create_unit_cube(MPI.COMM_WORLD, 10, 10, 10))
my_model.mesh = mesh

mat = F.Material(D_0=1, E_D=0)

volume = F.VolumeSubdomain(id=1, material=mat)
top_surface = F.SurfaceSubdomain(id=1, locator=lambda x: np.isclose(x[2], 1.0))
bottom_surface = F.SurfaceSubdomain(id=2, locator=lambda x: np.isclose(x[2], 0.0))
my_model.subdomains = [top_surface, bottom_surface, volume]

H = F.Species("H")
my_model.species = [H]

my_model.boundary_conditions = [
    F.FixedConcentrationBC(subdomain=top_surface, value=1.0, species=H),
    F.FixedConcentrationBC(subdomain=bottom_surface, value=0.0, species=H),
]

my_model.settings = F.Settings(atol=1e-10, rtol=1e-10, transient=False)

my_model.initialise()
my_model.run()

> ```

Reviewer 3: We could add the temperature field in this example to see it in action. Maybe create a new model object too with a different name to distinguish it

```{code-cell} ipython3
:tags: [hide-input]

from dolfinx import plot
import pyvista

pyvista.start_xvfb()
pyvista.set_jupyter_backend("html")

c = H.solution

topology, cell_types, geometry = plot.vtk_mesh(c.function_space)
u_grid = pyvista.UnstructuredGrid(topology, cell_types, geometry)
u_grid.point_data["c"] = c.x.array.real
u_grid.set_active_scalars("c")
u_plotter = pyvista.Plotter()
u_plotter.add_mesh(u_grid, cmap="viridis", show_edges=False)
u_plotter.add_mesh(u_grid, style="wireframe", color="white", opacity=0.2)

u_plotter.view_xy()

if not pyvista.OFF_SCREEN:
    u_plotter.show()
else:
    figure = u_plotter.screenshot("concentration.png")

> ```

Reviewer 3: Seems like we're looking at the concentration field, not the temperature

Reviewer 3: We could have both the temperature field and the concentration to see the influence

Reviewer 3: Plus we should either see how to have the default view be at an angle to see the whole cube, or just move to a 2D example

## Defining temperature as a FEniCS function ##

> Users can define temperature as a Dolfinx `Function`, which can be helpful when implementing multiphysics simulations (such as in OpenFOAM or OpenMC). For example, we want to have temperature function defined as:

Reviewer 3: ```suggestion
Users can define temperature as a dolfinx `fem.Function`, which can be helpful when implementing multiphysics simulations (such as in OpenFOAM or OpenMC). For example, we want to have a temperature function defined as:
```

Reviewer 3: I would, instead of saying defining temperature as a `fem.Function`, maybe say, we could say you can pass a `fem.Function` object to the temperature. Plus, maybe we don't need to mention multi-physics here, we could just mention how results from other programs can be converted into dolfinx functions
